rust/frontend/src/ty.rs

```diff
@@ -0,0 +1,163 @@
+//! This module implements the required conversions from Rust types to TVM types.
+//!
+//! In TVM frontend only conversions from Rust's 32-bits (POD) numeric types (i32, u32, f32)
+//! and 64-bits pointers are supported.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    ops::{Deref, DerefMut},
+};
+
+use ts;
+
+use Function;
+use Module;
+use NDArray;
+use TVMByteArray;
+use TVMContext;
+use TVMDeviceType;
+use TVMTypeCode;
+
+macro_rules! impl_prim_type {
+    ($type:ty, $variant:ident) => {
+        impl From<$type> for TVMTypeCode {
+            fn from(_arg: $type) -> Self {
+                TVMTypeCode::$variant
+            }
+        }
+
+        impl<'a> From<&'a $type> for TVMTypeCode {
+            fn from(_arg: &$type) -> Self {
+                TVMTypeCode::$variant
+            }
+        }
+
+        impl<'a> From<&'a mut $type> for TVMTypeCode {
+            fn from(_arg: &mut $type) -> Self {
+                TVMTypeCode::$variant
+            }
+        }
+    };
+}
+
+impl_prim_type!(TVMDeviceType, kDLInt);
+impl_prim_type!(TVMContext, kTVMContext);
+impl_prim_type!(TVMType, kTVMType);
+impl_prim_type!(Function, kFuncHandle);
+impl_prim_type!(Module, kModuleHandle);
+impl_prim_type!(NDArray, kArrayHandle);
+impl_prim_type!(TVMByteArray, kBytes);
+
+/// See the [module-level documentation](../ty/index.html) for more details.
+///
+/// Wrapper around underlying TVMType
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct TVMType {
+    // inner fields are (code: u8, bits: u8, lanes: u16)
+    pub inner: ts::TVMType,
+}
+
+impl TVMType {
+    pub(crate) fn new(type_code: u8, bits: u8, lanes: u16) -> Self {
+        TVMType {
+            inner: ts::TVMType {
+                code: type_code,
+                bits: bits,
+                lanes: lanes,
+            },
+        }
+    }
+}
+
+impl<'a> From<&'a str> for TVMType {
+    fn from(type_str: &'a str) -> Self {
+        match type_str {
+            "int" => TVMType::new(0, 32, 1),
+            "uint" => TVMType::new(1, 32, 1),
+            "float" => TVMType::new(2, 32, 1),
+            "handle" => TVMType::new(4, 64, 1),
+            _ => panic!("Unsupported type {:?}", type_str),
+        }
+    }
+}
+
+impl Display for TVMType {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
```

Comment:
Why not use [the existing convention](https://github.com/dmlc/tvm/blob/0e8d8fdd1be998460ea576dae9bcb7ba9530b7b1/include/tvm/runtime/packed_func.h#L946) of `{dtype}{bits}x{lanes}`?

Reply:
[java](https://github.com/dmlc/tvm/blob/master/jvm/core/src/main/java/ml/dmlc/tvm/TVMType.java#L78) and [python](https://github.com/dmlc/tvm/blob/master/python/tvm/_ffi/runtime_ctypes.py#L82) just concatenate the strings though

Reply:
right, but they produce strings like int8x32 or float32x1 or bool. this Display returns only {int, uint, float, and handle}